File: kNN/mykNN.m
function label = mykNN(x, complete_set, k, h)
  % mykNN: binary classifier: label is {-1, 1}
  % x: point to be classified
  % complete_set: training set with class labels
  % k: number of nearest neighbors
  % h: kernel bandwidth
  % label: class label of x

  % number of training samples
  training_set_point_qtty = size(complete_set, 1);

  % number of features
  feature_qtty = size(complete_set, 2) - 1;

  % initialize distances vector
  distances = zeros(training_set_point_qtty, 1);

  % compute distances
  for i = 1:training_set_point_qtty
    distances(i) = euclidean_distance(x, complete_set(i, 1:feature_qtty));
  end

  % sort distances
  [sorted_distances, sorted_indexes] = sort(distances);

  % initialize class labels vector
  class_labels = zeros(k, 1);

  % get class labels of k nearest neighbors
  for i = 1:k
    class_labels(i) = complete_set(sorted_indexes(i), feature_qtty + 1);
  end

  % compute kernel weights
  weights = zeros(k, 1);

  for i = 1:k
    weights(i) = kernel_function(sorted_distances(i), h);
  end

  % compute class label
  label = sign(sum(weights .* class_labels));

end

function distance = euclidean_distance(x, y)
  % euclidean_distance: computes the euclidean distance between two points
  % x, y: points
  % distance: euclidean distance between x and y

  distance = sqrt(sum((x - y) .^ 2));

end

function value = kernel_function(distance, h)
  % kernel_function: computes the kernel value using a normal distribution
  % distance: distance between points
  % h: kernel bandwidth
  % value: kernel value

  value = normpdf(distance, 0, h);

end

File: python/kNN/mykNN.py
import numpy as np
from scipy.spatial.distance import cdist

def mykNN_batch(Xt: np.ndarray, X: np.ndarray, Y: np.ndarray, k: int, h: float) -> np.ndarray:
  """
    Weighted k-nearest neighbors classifier
    Xt: target data, column vector of row vectors (samples, features)
    X: known data, column vector of row vectors (samples, features)
    Y: known labels, raw vector of integer labels (samples) expected labels are -1 and 1
    k: number of neighbors
    h: covariance factor
    returns
      Yt: predicted labels, raw vector of integer labels (samples) -1 or 1
    Weights are given by the multivariate normal distribution pdf
    with mean at the neighbor and bandwidth h
  """

  distances_sq = cdist(Xt, X, 'sqeuclidean')
  
  # Find indices of the k nearest neighbors for each target point
  indices = np.argpartition(distances_sq, k, axis=1)[:, :k]
  
  # Extract squared distances of the k neighbors using advanced indexing
  rows = np.arange(Xt.shape[0])[:, np.newaxis]
  knn_dist_sq = distances_sq[rows, indices]
  
  # Calculate weights using the kernel (multivariate normal PDF without normalization)
  weights = np.exp(-knn_dist_sq / (2 * h ** 2))
  
  # Retrieve the labels of the k nearest neighbors
  knn_labels = Y[indices]
  
  # Compute the weighted sum of labels
  weighted_sums = np.sum(weights * knn_labels, axis=1)
  
  # Determine the predicted labels based on the sign of the weighted sum
  Yt = np.where(weighted_sums >= 0, 1, -1)
  
  return Yt